File: python-uv-backend/scripts/setup_precommit.py
#!/usr/bin/env python3
"""
Set up pre-commit hooks for Python + uv projects.
"""

import argparse
import subprocess
import sys
from pathlib import Path


def run_command(cmd: list[str]) -> tuple[bool, str]:
    """Run a command and return success status and output."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            check=False
        )
        output = result.stdout + result.stderr
        return result.returncode == 0, output
    except FileNotFoundError:
        return False, f"Command not found: {' '.join(cmd)}"


def check_precommit_config_exists() -> bool:
    """Check if .pre-commit-config.yaml exists."""
    return Path(".pre-commit-config.yaml").exists()


def install_precommit() -> bool:
    """Install pre-commit using uv."""
    print("Installing pre-commit...")
    success, output = run_command(["uv", "add", "--dev", "pre-commit"])

    if not success:
        print(f"Failed to install pre-commit:\n{output}")
        return False

    print("✓ pre-commit installed")
    return True


def install_hooks() -> bool:
    """Install pre-commit hooks."""
    print("Installing pre-commit hooks...")
    success, output = run_command(["uv", "run", "pre-commit", "install"])

    if not success:
        print(f"Failed to install hooks:\n{output}")
        return False

    print("✓ pre-commit hooks installed")
    return True


def run_hooks_on_all_files() -> bool:
    """Run pre-commit on all files."""
    print("Running pre-commit on all files...")
    success, output = run_command(["uv", "run", "pre-commit", "run", "--all-files"])

    print(output)

    if not success:
        print("⚠ Some hooks failed on existing files (this is normal for first run)")
        print("Files have been auto-fixed where possible. Review and commit changes.")
        return True  # Not a fatal error

    print("✓ All hooks passed")
    return True


def main():
    parser = argparse.ArgumentParser(
        description="Set up pre-commit hooks for Python + uv project"
    )
    parser.add_argument(
        "--skip-run",
        action="store_true",
        help="Skip running hooks on all files after setup"
    )

    args = parser.parse_args()

    print("🔧 Setting up pre-commit hooks\n")

    # Check if config exists
    if not check_precommit_config_exists():
        print("ERROR: .pre-commit-config.yaml not found")
        print("Please create it first (use the template from assets/)")
        sys.exit(1)

    # Install pre-commit
    if not install_precommit():
        sys.exit(1)

    # Install hooks
    if not install_hooks():
        sys.exit(1)

    # Run on all files
    if not args.skip_run:
        if not run_hooks_on_all_files():
            sys.exit(1)

    print("\n✅ Pre-commit setup complete!")
    print("\nPre-commit will now run automatically on git commit.")
    print("To run manually: uv run pre-commit run --all-files")


if __name__ == "__main__":
    main()

File: frontend-plugin/skills/bun-vite-react/scripts/setup-precommit.ts
#!/usr/bin/env bun
/**
 * Set up pre-commit hooks for Bun + Vite + React + TypeScript projects.
 */

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[91m',
  green: '\x1b[92m',
  yellow: '\x1b[93m',
};

async function runCommand(cmd: string[]): Promise<{ success: boolean; output: string }> {
  try {
    const proc = Bun.spawn(cmd, {
      stdout: 'pipe',
      stderr: 'pipe',
    });

    const output = await new Response(proc.stdout).text();
    const error = await new Response(proc.stderr).text();
    const exitCode = await proc.exited;

    return {
      success: exitCode === 0,
      output: output + error,
    };
  } catch (error) {
    return { success: false, output: String(error) };
  }
}

async function checkPrecommitConfigExists(): Promise<boolean> {
  const file = Bun.file('.husky/pre-commit');
  return await file.exists();
}

async function installHusky(): Promise<boolean> {
  console.log('Installing husky...');
  const { success, output } = await runCommand(['bun', 'add', '-d', 'husky']);

  if (!success) {
    console.log(`${colors.red}Failed to install husky:${colors.reset}\n${output}`);
    return false;
  }

  console.log(`${colors.green}✓ husky installed${colors.reset}`);
  return true;
}

async function initHusky(): Promise<boolean> {
  console.log('Initializing husky...');
  const { success, output } = await runCommand(['bun', 'run', 'husky', 'init']);

  if (!success) {
    console.log(`${colors.red}Failed to initialize husky:${colors.reset}\n${output}`);
    return false;
  }

  console.log(`${colors.green}✓ husky initialized${colors.reset}`);
  return true;
}

async function setupPreCommitHook(): Promise<boolean> {
  console.log('Setting up pre-commit hook...');
  
  const hookContent = `#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

# Run verification before commit
bun run verify
`;

  try {
    await Bun.write('.husky/pre-commit', hookContent);
    await runCommand(['chmod', '+x', '.husky/pre-commit']);
    console.log(`${colors.green}✓ pre-commit hook configured${colors.reset}`);
    return true;
  } catch (error) {
    console.log(`${colors.red}Failed to create pre-commit hook:${colors.reset}\n${error}`);
    return false;
  }
}

async function addVerifyScript(): Promise<boolean> {
  console.log('Checking package.json for verify script...');
  
  try {
    const packageJson = await Bun.file('package.json').json();
    
    if (!packageJson.scripts) {
      packageJson.scripts = {};
    }

    if (!packageJson.scripts.verify) {
      packageJson.scripts.verify = 'bun run scripts/verify.ts';
      await Bun.write('package.json', JSON.stringify(packageJson, null, 2) + '\n');
      console.log(`${colors.green}✓ Added verify script to package.json${colors.reset}`);
    } else {
      console.log(`${colors.yellow}⚠ verify script already exists in package.json${colors.reset}`);
    }
    
    return true;
  } catch (error) {
    console.log(`${colors.yellow}⚠ Could not update package.json:${colors.reset} ${error}`);
    return true; // Non-fatal
  }
}

function parseArgs(): { skipRun: boolean } {
  const args = process.argv.slice(2);
  return {
    skipRun: args.includes('--skip-run'),
  };
}

async function main() {
  const args = parseArgs();

  console.log(`${colors.bright}🔧 Setting up pre-commit hooks${colors.reset}\n`);

  // Install husky
  if (!await installHusky()) {
    process.exit(1);
  }

  // Initialize husky
  if (!await initHusky()) {
    process.exit(1);
  }

  // Set up pre-commit hook
  if (!await setupPreCommitHook()) {
    process.exit(1);
  }

  // Add verify script to package.json
  await addVerifyScript();

  console.log(`\n${colors.green}✅ Pre-commit setup complete!${colors.reset}`);
  console.log('\nPre-commit hooks will now run automatically on git commit.');
  console.log('To run manually: bun run verify');
}

main();
